### ISSUE TITLE
question about the first training stage

### ISSUE THREAD

**Issue by colinzhaoxp** [skleee/GLEN#2]
hello authors, thanks for your open-source code, and I have learned a lot.

I have a question about the first stage of training pipeline:
https://github.com/skleee/GLEN/blob/d165ef8a59f6c09e9cbc17a0a454284bf09303c2/src/tevatron/modeling/glen_phase1.py#L88-L108
According formula 4 in the paper, the sequence-to-sequence cross-entropy loss should use the teacher-forcing style training, but the code shown above demonstrate the model generate docid token by token and compute loss (without teacher-forcing).

Are there any considerations for this type of training?

**Comment by skleee** [comment]
Hi, thank you for your interest in our work!

For a quick answer, **both phases 1 and 2 use the document representation as input**, token by token, during decoding **without teacher forcing**. This means we run a for-loop over each token in the sequence during generation.

As mentioned in Section 3.2 in the paper, we feed the document representation d_t, not z_t, into the decoder at each step.

<img width="545" height="314" alt="Image" src="https://github.com/user-attachments/assets/4f65ddcb-e6f6-4080-ae28-8b9a34d99c5c" />

This design choice was made to keep consistent the training procedures of phases 1 and 2, especially considering that the document identifiers may fluctuate in phase 2 and the model needs to remain stable under such conditions.

Please feel free to reach out if you have any further questions!

**Comment by colinzhaoxp** [comment]
thanks for your response!

I have the question about dynamic docid. 
In the first stage, the docid is generated by the key word extractor (BM25), then the model will learn the relationship between doc (and query) to the docid.  So the docids are fixed, such as Figure 1, the docid is `olymic-games-list`.
In the second stage, I guess the docid of the same document in Figure 1 is also `olymic-games-list`, so where is the dynamic docid ? they are same and not changed.

So my question is whether the docid will change?

**Comment by skleee** [comment]
@colinzhaoxp  Apologies for the delayed response. Thank you for asking such an insightful question! It gets to a very interesting part of our work.

**1. Will the docid change?**
To answer your question directly: **Yes, the DocID does change.**

You are correct that Figure 1 in the paper doesn't fully illustrate this due to space constraints. However, once the model completes phase 2 of training, it learns to adjust the initial IDs. This process is what creates the dynamic IDs.

We actually have a figure in our [poster](https://github.com/skleee/GLEN/blob/main/assets/glen_poster.pdf
) that visualizes this concept more clearly as below:

<img width="2574" height="1474" alt="Image" src="https://github.com/user-attachments/assets/3dd14b83-1acf-4c0c-acdd-51db8b3bf38d" />

Furthermore, we provide checkpoints trained on [NQ320k](https://drive.google.com/drive/folders/1ERopkRAJf7Ea-r_nJWoeaZFUp7e54eok?usp=sharing), [MS MARCO](https://drive.google.com/drive/folders/1mp4HIIbKnohNizLccaNFkJVMS-pJl_6T?usp=sharing). If you run inference using these checkpoints, you will notice that the outputs are different from those of a naive T5-base. Those are direct results of the learned dynamic document identifiers.


**2. Is zt equal to the one in the first stage (got by BM25)?**
There is no guarantee that they will be the same. The identifier z_t is what the model learns and predicts, so it can and will change from the initial BM25-generated ID based on the training.

I hope this clears things up. Thanks again for the great questions, and please let me know if you have any more!
